financial-templates-lib/price-feed/BalancerPriceFeed.js

```diff
@@ -0,0 +1,79 @@
+const assert = require("assert");
+const { PriceFeedInterface } = require("./PriceFeedInterface");
+const { BlockHistory, PriceHistory } = require("./utils");
+
+// Gets balancer prices as well as historical prices
+class BalancerPriceFeed extends PriceFeedInterface {
+  constructor(logger, web3, getTime, abi, address, tokenIn, tokenOut, lookback) {
+    assert(tokenIn, "BalancerPriceFeed requires tokenIn");
+    assert(tokenOut, "BalancerPriceFeed requires tokenOut");
+    assert(lookback, "BalancerPriceFeed requires lookback");
+    super();
+    this.logger = logger;
+    this.web3 = web3;
+    this.getTime = getTime;
+
+    this.contract = new web3.eth.Contract(abi, address);
+    this.currentPrice = null;
+    this.lastUpdateTime = null;
+    this.tokenIn = tokenIn;
+    this.tokenOut = tokenOut;
+    this.lookback = lookback;
+    // Provide a getblock function which returns the latest value if no number provided.
+    this.blockHistory = BlockHistory(number => web3.eth.getBlock(number >= 0 ? number : "latest"));
+
+    // Add a callback to get price, if an error is thrown returns undefined
+    this.priceHistory = PriceHistory(async number => {
+      try {
+        const result = await this.contract.methods.getSpotPriceSansFee(this.tokenIn, this.tokenOut).call(number);
+        return result;
+      } catch (err) {
+        this.logger.warn({
+          at: "BalancerPriceFeed.getPrice",
+          message: err.message,
+          number
+        });
+      }
+    });
+  }
+  getHistoricalPrice(time) {
+    // normally would bubble up errors, but this is not supposed to throw
+    try {
+      const block = this.blockHistory.getClosestTime(time);
+      return this.priceHistory.get(block.timestamp);
+    } catch (err) {
```

Comment:
Under what cases will these error? Could they throw for other reasons (unrelated to data availability)? I think these methods are allowed to throw -- I think they just aren't expected to throw if the issue is that the data isn't available since that _could_ be expected in certain situations -- or at least I think that was the intention in the uniswap price feed. Or maybe we don't expect data to not be available in the happy path of a functioning bot? I don't think we have a cohesive error-handling strategy across our components. It probably makes sense to think through how these errors should be propagated. We could:

1. Expect all errors to be handled by the callers and always return null.
2. Expect _certain_ errors to be handled by the callers and return `null` in those cases, but allow other errors to be propagated.
3. Always throw and expect the caller to catch if they want to gracefully handle the errors. This might require us to be careful around using Promise.all and prefer Promise.allSettled to ensure some failures don't block the expected timely resolution of other promises.

Thoughts?

Reply:
>  I don't think we have a cohesive error-handling strategy across our components. It probably makes sense to think through how these errors should be propagated
Yea I personally don't know what is going on around error handling currently, so feel safer not throwing and continuing, knowing a minor error can wreak havoc on the bots. Obviously though I don't want to introduce negative side effects, but currently its hard to reason about not knowing the API spec fully, so im really just guessing. 

I personally like throwing errors and avoid catching them until they have bubbled up as high in the app as they can reasonably go. I like to reason about them in this way: if an error is thrown, no matter what it is, that means whatever i was trying to do did not get done. This can get complicated under certain situations, like when using a promise.all, or partial state changes, so there are design implications, but its a trade off for easier to reason about errors. If you do this you need a strategy for handling the errors so that the app doesn't crash, but if you ensure all errors bubble up to the top, you typically only need to do this in 1 spot vs the patchwork of try catches currently.

I would be really happy to just let all of what i consider errors (maximal fail fast) to be thrown and handled at a higher level without having to know anything about the error that happened. does this sound reasonable?

Reply:
after looking at this more, it will throw if there is no price or block at that timestamp. There isnt any fetching going on.  So the question is really for this particular function, if the history doesnt exist, what should it do? right now its returning null